### ISSUE TITLE
Build failure of Mbed TLS 1.3 and 2.1 with GCC 7

### ISSUE THREAD

**Issue by gilles-peskine-arm** [ARMmbed/mbedtls#1188]
mbedtls-1.3 and mbedtls-2.1 do not build out of the box with `gcc-7 -Werror` on Linux.

```
$ git status
HEAD detached at mbedtls-2.1.9
nothing to commit, working tree clean
$ make lib
…
  CC    x509_crl.c
x509_crl.c: In function 'mbedtls_x509_crl_parse':
x509_crl.c:508:5: warning: this 'else' clause does not guard... [-Wmisleading-indentation]
     else
     ^~~~
x509_crl.c:514:9: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'else'
         if( ret == 0 )
         ^~
  CC    x509_crt.c
x509_crt.c: In function 'mbedtls_x509_crt_parse_path':
x509_crt.c:1186:62: warning: '%s' directive output may be truncated writing up to 255 bytes into a region of size 254 [-Wformat-truncation=]
         mbedtls_snprintf( entry_name, sizeof entry_name, "%s/%s", path, entry->d_name );
                                                              ^~
In file included from /usr/include/stdio.h:862:0,
                 from ../include/mbedtls/bignum.h:36,
                 from ../include/mbedtls/asn1.h:35,
                 from ../include/mbedtls/x509.h:32,
                 from ../include/mbedtls/x509_crt.h:32,
                 from x509_crt.c:40:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:64:10: note: '__builtin___snprintf_chk' output 2 or more bytes (assuming 257) into a destination of size 255
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
…
$ gcc --version 
gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0
…
```

The indentation issue already appeared with GCC 6 and there is a fix available in #1187 . It is specific to 2.1.

The diagnostic about the truncation of the `snprintf` output also appears in 1.3, but not in development. The value of the diagnostic in this case is dubious, but we should fix the broken build anyway. 

Looking at the code, the offending line is:

        mbedtls_snprintf( entry_name, sizeof entry_name, "%s/%s", path, entry->d_name );

It happens that `entry_name` is smaller than `d_name` on Linux. Just by one byte, so a potential fix is to make `entry_name` one byte larger. But the concatenation with `path` and a slash could be larger anyway, so the result may be truncated even with a larger `entry_name`. The development branch is saved by having a larger size by default but even then it could result in truncation. We should check the return value of `snprintf` (which would silence gcc's warning message) and report the error properly.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-1936

**Comment by sbutcher-arm** [comment]
This has been fixed in branch `mbedtls-2.1` with the merge of #1611.
